File: packages/fastapi/autoblock/rules_watcher.py
"""RulesWatcher — hot-reloads dynamic rules from Redis into RateLimiter."""
from __future__ import annotations

import asyncio
import json
import logging
from typing import Callable

import redis.asyncio as aioredis

from .key_builder import KeyBuilder
from .types import RuleConfig, LimitsConfig, PenaltyConfig, PenaltyThresholds

log = logging.getLogger(__name__)

_DEFAULT_INTERVAL = 30  # seconds


class RulesWatcher:
    """
    Polls ``ab:{tenant}:rules:endpoint`` (Redis hash) every *interval_seconds*
    and hot-reloads rules into the associated :class:`~autoblock.RateLimiter`.

    Usage::

        watcher = RulesWatcher(redis, limiter)
        await watcher.start()        # fires once immediately, then every 30 s
        # later…
        await watcher.stop()

    Or as an async context manager::

        async with RulesWatcher(redis, limiter):
            await app()
    """

    def __init__(
        self,
        redis: aioredis.Redis,
        limiter,  # RateLimiter (avoid circular import)
        *,
        interval_seconds: int = _DEFAULT_INTERVAL,
        on_reload: Callable[[list[RuleConfig]], None] | None = None,
        on_error: Callable[[Exception], None] | None = None,
    ) -> None:
        self._redis = redis
        self._limiter = limiter
        self._interval = interval_seconds
        self._on_reload = on_reload
        self._on_error = on_error
        self._keys = KeyBuilder(limiter._config.tenant)
        self._task: asyncio.Task | None = None

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def start(self) -> None:
        """Start background polling (idempotent)."""
        if self._task is not None:
            return
        await self.poll()  # immediate first load
        self._task = asyncio.create_task(self._loop(), name="autoblock-rules-watcher")

    async def stop(self) -> None:
        """Cancel background polling."""
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    async def __aenter__(self) -> "RulesWatcher":
        await self.start()
        return self

    async def __aexit__(self, *_) -> None:
        await self.stop()

    # ------------------------------------------------------------------
    # Core poll
    # ------------------------------------------------------------------

    async def poll(self) -> list[RuleConfig]:
        """Single poll cycle — callable directly in tests."""
        raw: dict[str, str] = await self._redis.hgetall(self._keys.rules())
        dynamic = _parse_dynamic_rules(raw)
        self._limiter.merge_rules(dynamic)
        if self._on_reload:
            self._on_reload(dynamic)
        return dynamic

    # ------------------------------------------------------------------
    # Background loop (ticker + pub/sub push)
    # ------------------------------------------------------------------

    async def _loop(self) -> None:
        # Subscribe for push-based invalidation from the engine.
        push_task = asyncio.create_task(
            self._subscribe_push(), name="autoblock-rules-push"
        )
        try:
            while True:
                await asyncio.sleep(self._interval)
                try:
                    await self.poll()
                except asyncio.CancelledError:
                    raise
                except Exception as exc:
                    log.warning("autoblock RulesWatcher poll error: %s", exc)
                    if self._on_error:
                        self._on_error(exc)
        finally:
            push_task.cancel()
            try:
                await push_task
            except (asyncio.CancelledError, Exception):
                pass

    async def _subscribe_push(self) -> None:
        """Subscribe to ab:{tenant}:rules:changed and reload immediately on message."""
        try:
            async with self._redis.pubsub() as pubsub:
                await pubsub.subscribe(self._keys.rules_changed())
                async for message in pubsub.listen():
                    if message.get("type") == "message":
                        log.debug("autoblock RulesWatcher: push notification received")
                        try:
                            await self.poll()
                        except Exception as exc:
                            log.warning("autoblock RulesWatcher push-reload error: %s", exc)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            # Pub/sub unavailable — ticker-only fallback is still running.
            log.debug("autoblock RulesWatcher: pub/sub unavailable (%s), poll-only mode", exc)


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _parse_dynamic_rules(raw: dict[str, str]) -> list[RuleConfig]:
    rules: list[RuleConfig] = []
    for json_str in raw.values():
        try:
            r = json.loads(json_str)
            if r.get("enabled") is False:
                continue
            rules.append(_map_redis_rule(r))
        except Exception:
            pass  # skip malformed entries
    return rules


def _map_redis_rule(r: dict) -> RuleConfig:
    return RuleConfig(
        id=r.get("id", ""),
        enabled=r.get("enabled", True),
        description=r.get("description", ""),
        dimensions=r.get("dimensions", ["ip"]),
        endpoint_pattern=r.get("path") or r.get("endpoint_pattern", ""),
        methods=r.get("methods", ["*"]),
        algorithm=r.get("algorithm", "hybrid"),
        limits=LimitsConfig(
            requests=r.get("limit", 100),
            window_seconds=r.get("window_seconds", 60),
            burst=r.get("burst") or r.get("limit", 100),
        ),
        penalties=_map_penalties(r.get("penalties", {})),
    )


def _map_penalties(p: dict) -> PenaltyConfig:
    if not p:
        return PenaltyConfig()
    raw_thresholds = p.get("thresholds")
    thresholds = (
        PenaltyThresholds(**raw_thresholds)
        if isinstance(raw_thresholds, dict)
        else PenaltyThresholds()
    )
    return PenaltyConfig(
        warn=p.get("warn"),
        slow=p.get("slow"),
        block=p.get("block"),
        blacklist=p.get("blacklist"),
        thresholds=thresholds,
        ttl_seconds=p.get("ttl_seconds", 86400),
    )

File: packages/core/src/rules-watcher.ts
import type { Redis } from 'ioredis'
import type { RateLimiter } from './rate-limiter.js'
import { KeyBuilder } from './key-builder.js'
import type { RuleConfig } from './types.js'

export interface RulesWatcherOptions {
  /** Poll interval in milliseconds. Default: 30 000 (30 s). */
  intervalMs?: number
  /** Called after each successful reload with the new dynamic rules. */
  onReload?: (rules: RuleConfig[]) => void
  /** Called when a Redis error occurs during polling. */
  onError?: (err: Error) => void
}

/**
 * RulesWatcher polls `ab:{tenant}:rules:endpoint` (Redis hash) every 30 s
 * and hot-reloads rate-limit rules into the associated RateLimiter without
 * requiring a service restart.
 *
 * Usage:
 *   const watcher = new RulesWatcher(redis, limiter)
 *   watcher.start()          // polls immediately, then every 30 s
 *   // later...
 *   watcher.stop()
 */
export class RulesWatcher {
  private timer?: ReturnType<typeof setInterval>
  private subscriber?: Redis
  private readonly keys: KeyBuilder

  constructor(
    private readonly redis: Redis,
    private readonly limiter: RateLimiter,
    private readonly opts: RulesWatcherOptions = {},
  ) {
    this.keys = new KeyBuilder(limiter.tenant)
  }

  /**
   * Start polling + pub/sub subscription. Idempotent.
   * An immediate poll fires before the first interval.
   * Pub/sub failures degrade gracefully to ticker-only mode.
   */
  start(): void {
    if (this.timer !== undefined) return // already running
    const interval = this.opts.intervalMs ?? 30_000

    const tick = (): void => {
      this.poll().catch(err => this.opts.onError?.(err as Error))
    }

    tick()
    this.timer = setInterval(tick, interval)
    if (typeof this.timer.unref === 'function') this.timer.unref()

    // Subscribe for push-based invalidation (best-effort).
    this.startSubscriber().catch(() => {
      // Pub/sub unavailable — ticker-only fallback already running.
    })
  }

  /** Stop polling and unsubscribe. */
  stop(): void {
    if (this.timer !== undefined) {
      clearInterval(this.timer)
      this.timer = undefined
    }
    if (this.subscriber !== undefined) {
      this.subscriber.disconnect()
      this.subscriber = undefined
    }
  }

  /** Perform a single poll cycle (also callable manually in tests). */
  async poll(): Promise<RuleConfig[]> {
    const raw = await this.redis.hgetall(this.keys.rules())
    const dynamic = parseDynamicRules(raw ?? {})
    this.limiter.mergeRules(dynamic)
    this.opts.onReload?.(dynamic)
    return dynamic
  }

  // ---- Pub/sub subscription ------------------------------------------------

  private async startSubscriber(): Promise<void> {
    // ioredis requires a dedicated connection for subscribe mode.
    this.subscriber = this.redis.duplicate()
    await this.subscriber.subscribe(this.keys.rulesChanged())

    this.subscriber.on('message', (_channel: string, _message: string) => {
      // Engine published a change — reload immediately.
      this.poll().catch(err => this.opts.onError?.(err as Error))
    })

    this.subscriber.on('error', () => {
      // Silently ignore — ticker fallback is still running.
    })
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function parseDynamicRules(raw: Record<string, string>): RuleConfig[] {
  const rules: RuleConfig[] = []
  for (const json of Object.values(raw)) {
    try {
      const r = JSON.parse(json) as Record<string, unknown>
      if (r['enabled'] === false) continue
      rules.push(mapRedisRule(r))
    } catch {
      // skip malformed JSON — engine writes are atomic, so this is rare
    }
  }
  return rules
}

function mapRedisRule(r: Record<string, unknown>): RuleConfig {
  const penalties = (r['penalties'] as RuleConfig['penalties'] | undefined) ?? {}
  return {
    id:              String(r['id'] ?? ''),
    enabled:         (r['enabled'] as boolean | undefined) ?? true,
    description:     r['description'] as string | undefined,
    dimensions:      (r['dimensions'] as RuleConfig['dimensions'] | undefined) ?? ['ip'],
    endpointPattern: String(r['path'] ?? r['endpoint_pattern'] ?? ''),
    methods:         (r['methods'] as string[] | undefined) ?? ['*'],
    algorithm:       (r['algorithm'] as RuleConfig['algorithm'] | undefined) ?? 'hybrid',
    limits: {
      requests:      Number(r['limit'] ?? 100),
      windowSeconds: Number(r['window_seconds'] ?? 60),
      burst:         Number(r['burst'] ?? r['limit'] ?? 100),
    },
    penalties,
    tags: (r['tags'] as string[] | undefined) ?? [],
  }
}
